### PR TITLE
[Innovation] Onboarding: Tours & Hints

### PR DESCRIPTION
**Pull request by juanfra**
WIP: Onboarding - Tours & Hints.

### CODE REVIEW

**Comment by juanfra** [comment]
Adding abstract classes to make it easier to create Tours and Hints on plugins. [See example](https://github.com/the-events-calendar/event-tickets/pull/2015)

Next steps:

- Add methods to save the times it was displayed (and increase each time) so we can display them only on the first time, or for the first 5, etc.
- Try to find other places to abstract and see ways to make it easier to be extended from plugins.

**Comment by juanfra** [comment]
Abstract things so it's easier to use in our plugins. This is kind of ready for reviews.

File: src/Tribe/Onboarding/Main.php [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+		wp_localize_script( 'tribe-onboarding-js', 'TribeOnboardingTour', $data );
+	}
+
+	/**
+	 * Localize hints data.
+	 *
+	 * @since TBD
+	 *
+	 * @param string $hook The current admin page.
+	 * @return void
```
Comment (sc0ttkclark) [inline src/Tribe/Onboarding/Main.php:125]:
```suggestion
```

File: src/Tribe/Onboarding/Main.php [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+		return $data;
+	}
+
+	/**
+	 * Localize tour data.
+	 *
+	 * @since TBD
+	 *
+	 * @param string $hook The current admin page.
+	 * @return void
```
Comment (sc0ttkclark) [inline src/Tribe/Onboarding/Main.php:111]:
```suggestion
```

**Review by borkweb — DISMISSED** [review]
This is great work, @juanfra! I'm good with picking a release (how about `B21.eclair`?) and merging this in.

**Comment by juanfra** [comment]
Thanks @borkweb! Yes, absolutely

**Comment by juanfra** [comment]
@paulmskim I'm getting the following lint thing:

```
/home/runner/work/tribe-common/tribe-common/src/resources/js/onboarding.js
  39:3  error  Parsing error: The keyword 'const' is reserved
```

Am I missing something?

**Comment by paulmskim** [comment]
@juanfra no, this is a eslint configuration, i'll push a fix to your branch.

File: src/resources/postcss/onboarding.pcss [hunk 3]
```diff
@@ -0,0 +1,157 @@
+/*
+ * Onboarding CSS (Tours & Hints)
+ * -------------------------------------------------------------
+ * This file includes styles for the Onboarding Tours and Hints.
+ * These styles go on top of the ones from Intro.js
+ */
+
+/* Utilities */
+@import "utilities/_all.pcss";
```
Comment (paulmskim) [inline src/resources/postcss/onboarding.pcss:9]: I think you'll need to set the variables skeleton and full as dependencies of this pcss file and remove this, as we've moved onto preserving custom properties.

File: src/resources/postcss/onboarding.pcss [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+.introjs-tooltipbuttons {
+	border-top: none;
+}
+
+.introjs-bullets {
+	ul {
```
Comment (paulmskim) [inline src/resources/postcss/onboarding.pcss:16]: you can probably combine these lines

File: src/resources/postcss/onboarding.pcss [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+		border: 0;
+		text-shadow: none;
+
+		&:focus {
+			box-shadow: none;
+		}
+	}
+
+	.introjs-prevbutton {
+		color: #c9c9c9;
```
Comment (paulmskim) [inline src/resources/postcss/onboarding.pcss:64]: do we have a similar color variable we could probably use here?

File: src/resources/postcss/onboarding.pcss [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+		&,
+		&:hover {
+			background-color: transparent;
+			border-color: transparent;
+		}
+	}
+
+	.introjs-nextbutton {
+		background-color: #f7f6f6;
```
Comment (paulmskim) [inline src/resources/postcss/onboarding.pcss:74]: do we have a similar color variable we could probably use here?

File: src/Tribe/Service_Providers/Onboarding.php [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	 *
+	 * @since TBD
+	 */
+	public function register_assets() {
+		$main = \Tribe__Main::instance();
+
+		tribe_asset(
+			$main,
+			'intro-js',
+			'node_modules/intro.js/intro.js',
```
Comment (paulmskim) [inline src/Tribe/Service_Providers/Onboarding.php:47]: we need to either commit this to our repo or preserve this during the build process, as we remove the `node_modules` folder after the build.

File: src/Tribe/Service_Providers/Onboarding.php [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+			[
+				'groups'       => 'tribe-onboarding',
+				'conditionals' => [ $this, 'should_enqueue_assets' ],
+			]
+		);
+
+		tribe_asset(
+			$main,
+			'intro-styles',
+			'node_modules/intro.js/introjs.css',
```
Comment (paulmskim) [inline src/Tribe/Service_Providers/Onboarding.php:59]: we need to either commit this to our repo or preserve this during the build process, as we remove the `node_modules` folder after the build.

File: src/Tribe/Service_Providers/Onboarding.php [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+		$main = \Tribe__Main::instance();
+
+		tribe_asset(
+			$main,
+			'intro-js',
+			'node_modules/intro.js/intro.js',
+			[],
+			[ 'admin_enqueue_scripts' ],
+			[
+				'groups'       => 'tribe-onboarding',
```
Comment (paulmskim) [inline src/Tribe/Service_Providers/Onboarding.php:51]: let's create a static var for this group name

**Review by paulmskim — CHANGES_REQUESTED** [review]
Lots of great stuff here, it's almost there!

File: src/Tribe/Service_Providers/Onboarding.php (hunk 7 above)
Comment (juanfra) [reply src/Tribe/Service_Providers/Onboarding.php:47]: Right! I forgot to add it to the whitelist, like we do with the `mt-a11y-dialog`. Good catch!

File: src/resources/postcss/onboarding.pcss (hunk 3 above)
Comment (juanfra) [reply src/resources/postcss/onboarding.pcss:9]: Perfecto! good catch! Yeah, seems like that's what we need to do. Worked on this before the split.

File: src/resources/postcss/onboarding.pcss (hunk 5 above)
Comment (juanfra) [reply src/resources/postcss/onboarding.pcss:64]: Will check that 👍

File: src/resources/postcss/onboarding.pcss (hunk 6 above)
Comment (juanfra) [reply src/resources/postcss/onboarding.pcss:74]: Will check that 👍

File: src/Tribe/Onboarding/Main.php [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+		/**
+		 * Filter the data we're using to localize the hints.
+		 *
+		 * Since TBD
+		 *
+		 * @param array $data An array with the hints data.
+		 *
+		 * @return array $data An array with the hints data.
+		 */
+		$data = apply_filters( 'tribe_onboarding_hints_data', [] );
```
[the remark this answers is not in the document]
Comment (juanfra) [reply src/Tribe/Onboarding/Main.php:100]: Exactly, like the docblock says! good catch. Probably a wrong copy/paste, or something 😁

**Comment by juanfra** [comment]
This one is late for `foret-noire`, need to find a release, probably `falcarius` (once things are synched, after `foret-noire` release).

**Review by paulmskim — DISMISSED** [review]
looks great! excited for this!

File: src/Tribe/Service_Providers/Onboarding.php [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+ * Handles the registration and creation of our async process handlers.
+ */
+class Onboarding extends \tad_DI52_ServiceProvider {
+
+	/**
+	 * The Onboarding assets group identifier.
+	 *
+	 * @var string
+	 */
+	private static $assets_group_id = 'tribe-onboarding';
```
Comment (paulmskim) [inline src/Tribe/Service_Providers/Onboarding.php:18]: let's make this public, we do that in other places. also, let's call it `$group_key`, mimicking the same in TEC.

**Review by paulmskim — APPROVED** [review]

**Pull request merged**